File: go/2021/aug/range-sum-query-immutable.go
package aug

// NumArray define
type NumArray struct {
	prefixSum []int
}

// Constructor303 define
func Constructor303(nums []int) NumArray {
	for i := 1; i < len(nums); i++ {
		nums[i] += nums[i-1]
	}
	return NumArray{prefixSum: nums}
}

// SumRange define
func (this *NumArray) SumRange(i int, j int) int {
	if i > 0 {
		return this.prefixSum[j] - this.prefixSum[i-1]
	}
	return this.prefixSum[j]
}

File: code/java/src/main/java/cn/az/code/year2021/aug/RangeSumQueryImmutable.java
package cn.az.code.year2021.aug;

/**
 * @author az
 * @since 2021-08-16
 */
public class RangeSumQueryImmutable {

    int[] nums;

    public RangeSumQueryImmutable(int[] nums) {
        for (int i = 1; i < nums.length; i++) nums[i] += nums[i - 1];

        this.nums = nums;
    }

    public int sumRange(int i, int j) {
        if (i == 0) return nums[j];

        return nums[j] - nums[i - 1];
    }
}
